pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,347 @@
+            progress_callback, count = map_out
+            result_data["files"] += count
+            progress_callback(map_in)  # hexdigest
+            return True
+
+        sorted_todo_hexdigests = sorted(
+            todo_hexdigests, key=lambda hexdigest: -snapshotter.hexdigest_to_snapshotfiles[hexdigest][0].file_size
+        )
+        if not parallel_map_to(
+            fun=_upload_files_in_thread, iterable=sorted_todo_hexdigests, result_callback=_result_cb, n=self.parallel
+        ):
+            progress.add_fail()
+
+        progress.done()
+
+        if progress.finished_successfully:
+            self.log.info("All delta file were submitted for uploading")
+        else:
+            raise BackupFailure("Error while submitting delta backup files for uploading")
+
+        return result_data["files"]
+
+    def _delta_upload(self, snapshot_result: SnapshotResult, snapshotter: Snapshotter, start_time_utc):
+        callback_queue = Queue()
+
+        total_stored_size = 0
+        total_size = 0
+
+        # Determine which digests already exist and which need to be uploaded, also restore the backup size of re-used
+        # files from manifests
+        snapshot_hashes = set(snapshot_result.hashes)
+        uploaded_hashes = set()
+        all_existing_files = self._list_existing_files()
+        for snapshot_file in snapshot_result.state.files:
+            if snapshot_file.hexdigest in all_existing_files:
+                snapshot_file_from_manifest = all_existing_files[snapshot_file.hexdigest]
+                total_stored_size += snapshot_file_from_manifest.stored_file_size
+                total_size += snapshot_file_from_manifest.file_size
+                uploaded_hashes.add(
+                    SnapshotHash(
+                        hexdigest=snapshot_file_from_manifest.hexdigest, size=snapshot_file_from_manifest.file_size
+                    )
+                )
+
+        todo = snapshot_hashes.difference(uploaded_hashes)
+        todo_count = len(todo)
+
+        self.log.info(
+            "Submitting delta hashes to upload: %r, total hashes in the snapshot: %r", todo_count, len(snapshot_hashes)
+        )
+
+        upload_files_count = self._submit_uploading(callback_queue=callback_queue, todo=todo, snapshotter=snapshotter)
+
+        processed_files_count = 0
+        process_futures = []
+        upload_size = 0
+        upload_stored_size = 0
+
+        self.log.info("Waiting for the upload of delta backup files (%r) to finish", upload_files_count)
+
+        def _post_process_delta_file(upload_result):
+            opaque = upload_result["opaque"]
+            storage_delta_path = os.path.join(self.site_config["prefix"], "basebackup_delta")
+            old_key = os.path.join(storage_delta_path, opaque["object_path"])
+            new_key = os.path.join(storage_delta_path, opaque["hexdigest"])
+            self.storage.copy_file(source_key=old_key, destination_key=new_key)
```

Comment:
Will this just overwrite existing file with the same name? Just checking there won't be problems if we did new partial delta backup and pghoard was shut down before that completed and we have a bunch of remote files that aren't tracked (probably not an issue from storage usage point-of-view but just need to make sure they don't interfere with the rest of the logic).

Reply:
After I moved this part into `transfer.py` the upload should retry until succeeds. After upload fails, file on the disk won't be removed and possible temp object with the same name will exist as well, but will be rewritten and re-copied with the next successful retry.